File: java/src/jan/week3/boj/보석도둑_박재환.java
package jan.week3.boj;

import java.util.*;
import java.io.*;

public class 보석도둑_박재환 {
    static BufferedReader br;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        init();
        br.close();
    }
    static StringTokenizer st;
    static int n, k;        // 보석, 가방
    static int[][] diamonds;
    static int[] bags;
    static void init() throws IOException {
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        diamonds = new int[n][2];
        for(int i=0; i<n; i++) {
            st = new StringTokenizer(br.readLine().trim());
            int weight = Integer.parseInt(st.nextToken());
            int price = Integer.parseInt(st.nextToken());

            diamonds[i] = new int[] {weight, price};
        }
        bags = new int[k];
        for(int i=0; i<k; i++) bags[i] = Integer.parseInt(br.readLine().trim());

        System.out.println(getMaxPrice());
    }

    static long getMaxPrice() {
        long totalPrice = 0;
        Arrays.sort(diamonds, (a, b) -> Integer.compare(a[0], b[0]));       // 보석들을 무게가 가벼운순으로 오름차순 정렬
        Arrays.sort(bags);

        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> Integer.compare(b, a));       // 가격을 PQ 에 저장 ( 가격이 높은 순으로 )
        int diamondIdx = 0;
        for(int bag : bags) {
            while(diamondIdx < n && bag >= diamonds[diamondIdx][0]) pq.offer(diamonds[diamondIdx++][1]);

            while(!pq.isEmpty()) {
                totalPrice += pq.poll();
                break;
            }
        }

        return totalPrice;
    }
}

File: python/src/jan/week3/boj/보석도둑_박재환.py
def solution():
    import sys
    input = sys.stdin.readline

    n, k = map(int, input().split())
    diamonds = [list(map(int, input().split())) for _ in range(n)]
    bags = []
    for _ in range(k):
        bags.append(int(input()))

    def get_max_price():
        nonlocal  diamonds, bags
        import heapq as hq
        h = []

        diamonds.sort(key=lambda x: x[0])   # 무게가 가벼운 순
        bags.sort()

        total_price = 0
        idx = 0
        for bag in bags:
            while idx < n and diamonds[idx][0] <= bag:
                hq.heappush(h, -diamonds[idx][1])
                idx += 1

            if h:
                total_price += -hq.heappop(h)

        return total_price

    return get_max_price()

if __name__ == "__main__":
    print(solution())

